rust/sleeper_core/src/datafusion/leaf_partition_query.rs

```diff
@@ -0,0 +1,218 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+use crate::{
+    CommonConfig, SleeperPartitionRegion,
+    datafusion::{
+        CompletionOptions, SleeperOperations,
+        output::CompletedOutput,
+        sketch::{Sketcher, output_sketch},
+        util::explain_plan,
+    },
+};
+#[cfg(doc)]
+use arrow::record_batch::RecordBatch;
+use datafusion::{common::plan_err, logical_expr::Expr, physical_plan::displayable};
+use datafusion::{
+    dataframe::DataFrame,
+    error::DataFusionError,
+    execution::{config::SessionConfig, context::SessionContext},
+};
+use log::info;
+use objectstore_ext::s3::ObjectStoreFactory;
+use std::{
+    fmt::{Display, Formatter},
+    sync::Arc,
+};
+
+/// All information needed for a Sleeper leaf partition query.
+#[derive(Debug, Default)]
+pub struct LeafPartitionQueryConfig<'a> {
+    /// Basic information
+    pub common: CommonConfig<'a>,
+    /// Query ranges
+    pub ranges: Vec<SleeperPartitionRegion<'a>>,
+    /// Should sketches be produced?
+    pub write_quantile_sketch: bool,
+    /// Should logical/physical plan explanation be logged?
+    pub explain_plans: bool,
+}
+
+impl Display for LeafPartitionQueryConfig<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Query config: {}, ranges: {:?} write quantile sketches: {}",
+            self.common, self.ranges, self.write_quantile_sketch
+        )
+    }
+}
+
+/// Manages and executes a Sleeper leaf partition query.
+#[derive(Debug)]
+pub struct LeafPartitionQuery<'a> {
+    /// The configuration information for the leaf query
+    config: &'a LeafPartitionQueryConfig<'a>,
+    /// Used to create object store implementations
+    store_factory: &'a ObjectStoreFactory,
+}
+
+impl<'a> LeafPartitionQuery<'a> {
+    pub fn new(
+        config: &'a LeafPartitionQueryConfig<'a>,
+        store_factory: &'a ObjectStoreFactory,
+    ) -> LeafPartitionQuery<'a> {
+        Self {
+            config,
+            store_factory,
+        }
+    }
+
+    /// Executes a Sleeper leaf partition query.
+    ///
+    /// The object store factory must be able to produce an [`object_store::ObjectStore`] capable of reading
+    /// from the input URLs and writing to the output URL (if writing results to a file).
+    ///
+    /// # Errors
+    /// There must be at least one query region specified.
+    pub async fn run_query(&self) -> Result<CompletedOutput, DataFusionError> {
```

Comment:
It doesn't look like the new code has any tests?

I can see the stacked PR for the tests of the CLI code, but that's just for validation of the command line arguments.